Python/Product/TestAdapter.Executor/Pytest/PytestExtensions.cs

```diff
@@ -0,0 +1,117 @@
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABILITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using Microsoft.VisualStudio.TestPlatform.ObjectModel;
+
+namespace Microsoft.PythonTools.TestAdapter.Pytest {
+    static class PyTestDiscoveryReader {
+
+        public static TestCase ToVsTestCase(this PytestTest test, string projectHome, Dictionary<string, PytestParent> parentMap) {
+            if (String.IsNullOrEmpty(projectHome)) {
+                throw new ArgumentException(nameof(projectHome));
+            }
+            if (parentMap == null) {
+                throw new ArgumentException(nameof(parentMap));
+            }
+
+            var sourceAndLineNum = test.Source.Replace(".\\", "");
+            String[] sourceParts = sourceAndLineNum.Split(':');
+            int line = 0;
+
+            if (sourceParts.Length != 2 ||
+                !Int32.TryParse(sourceParts[1], out line) ||
+                String.IsNullOrWhiteSpace(test.Name) ||
+                String.IsNullOrWhiteSpace(test.Id)) {
+                throw new FormatException("Invalid pytest test discovered: " + test.ToString());
+            }
+
+            //bschnurr todo: fix codepath for files outside of project
+            var fullSourcePathNormalized = Path.Combine(projectHome, sourceParts[0]).ToLowerInvariant();
+            var fullyQualifiedName = CreateFullyQualifiedTestNameFromId(test.Id);
+
+            var tc = new TestCase(fullyQualifiedName, PythonConstants.PytestExecutorUri, fullSourcePathNormalized) {
+                DisplayName = test.Name,
+                LineNumber = line,
+                CodeFilePath = fullSourcePathNormalized
+            };
+
+            tc.SetPropertyValue(Constants.PytestIdProperty, test.Id);
+            tc.SetPropertyValue(Constants.PyTestXmlClassNameProperty, CreateXmlClassName(test, parentMap));
+            tc.SetPropertyValue(Constants.PytestTestExecutionPathPropertery, GetAbsoluteTestExecutionPath(fullSourcePathNormalized, test.Id));
+            return tc;
+        }
+
+
+        /// <summary>
+        /// Creates a classname that matches the junit testresult generated one so that we can match testresults with testcases
+        /// Note if a function doesn't have a class, its classname appears to be the filename without an extension
+        /// </summary>
+        /// <param name="t"></param>
+        /// <param name="parentMap"></param>
+        /// <returns></returns>
+        public static string CreateXmlClassName(PytestTest t, Dictionary<string, PytestParent> parentMap) {
```

Comment:
these 3 static methods can be private it looks like

Reply:
the idea was to make it accessible for testing. I'll make it internal instead